### PR TITLE
Unit tests -> Do unit tests for PopularMoviesViewModel

### PR DESCRIPTION
**Pull request by Nataniel-Antosik**

### CODE REVIEW

**Comment by Nataniel-Antosik** [comment]
# Problem with test case
I have a problem with a test case that contains a case when `use case` returns throwable then method run method from navigator to show snackbar and also live data `isLoaderVisible` is changed from `false` to `true` (this case I have).

**Comment by Nataniel-Antosik** [comment]
## And all tested passed but I have this exception in terminal in android studio
```
Exception in thread "Test worker @coroutine#8" io.mockk.MockKException: no answer found for: PopularMoviesFragmentNavigator(#2).errorSnackBar(Something went wrong)
	at io.mockk.impl.stub.MockKStub.defaultAnswer(MockKStub.kt:93)
	at io.mockk.impl.stub.MockKStub.answer(MockKStub.kt:42)
	at io.mockk.impl.recording.states.AnsweringState.call(AnsweringState.kt:16)
	at io.mockk.impl.recording.CommonCallRecorder.call(CommonCallRecorder.kt:53)
	at io.mockk.impl.stub.MockKStub.handleInvocation(MockKStub.kt:266)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1.invocation(JvmMockFactoryHelper.kt:23)
	at io.mockk.proxy.jvm.advice.Interceptor.call(Interceptor.kt:21)
	at com.example.benchproject.app.common.navigator.FragmentNavigator.errorSnackBar(FragmentNavigator.kt:21)
	at com.example.benchproject.app.popular.movies.viewModel.PopularMoviesViewModel$getPopularMovie$1.invokeSuspend(PopularMoviesViewModel.kt:49)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
	at kotlinx.coroutines.BuildersKt.launch(Unknown Source)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source)
	at com.example.benchproject.app.popular.movies.viewModel.PopularMoviesViewModel.getPopularMovie(PopularMoviesViewModel.kt:32)
	at com.example.benchproject.app.popular.movies.viewModel.PopularMoviesViewModel.<init>(PopularMoviesViewModel.kt:29)
	at com.example.benchproject.app.popular.movies.viewModel.PopularMoviesViewModelTest$when method get throwable from use case, loader visibility should be true$1.invokeSuspend(PopularMoviesViewModelTest.kt:74)
	at com.example.benchproject.app.popular.movies.viewModel.PopularMoviesViewModelTest$when method get throwable from use case, loader visibility should be true$1.invoke(PopularMoviesViewModelTest.kt)
	at com.example.benchproject.app.popular.movies.viewModel.PopularMoviesViewModelTest$when method get throwable from use case, loader visibility should be true$1.invoke(PopularMoviesViewModelTest.kt)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTestCoroutine$2.invokeSuspend(TestBuilders.kt:208)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTestCoroutine$2.invoke(TestBuilders.kt)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTestCoroutine$2.invoke(TestBuilders.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startCoroutineUndispatched(Undispatched.kt:55)
	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:112)
	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTestCoroutine(TestBuilders.kt:207)
	at kotlinx.coroutines.test.TestBuildersKt.runTestCoroutine(Unknown Source)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$1$1.invokeSuspend(TestBuilders.kt:167)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$1$1.invoke(TestBuilders.kt)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$1$1.invoke(TestBuilders.kt)
	at kotlinx.coroutines.test.TestBuildersJvmKt$createTestResult$1.invokeSuspend(TestBuildersJvm.kt:13)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at kotlinx.coroutines.test.TestBuildersJvmKt.createTestResult(TestBuildersJvm.kt:12)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest(TestBuilders.kt:166)
	at kotlinx.coroutines.test.TestBuildersKt.runTest(Unknown Source)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest(TestBuilders.kt:154)
	at kotlinx.coroutines.test.TestBuildersKt.runTest(Unknown Source)
	at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest$default(TestBuilders.kt:147)
	at kotlinx.coroutines.test.TestBuildersKt.runTest$default(Unknown Source)
	at com.example.benchproject.app.popular.movies.viewModel.PopularMoviesViewModelTest.when method get throwable from use case, loader visibility should be true(PopularMoviesViewModelTest.kt:72)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [CoroutineId(8), "coroutine#8":StandaloneCoroutine{Cancelling}@26ca61bf, Dispatchers.Main]
BUILD SUCCESSFUL in 3s
35 actionable tasks: 4 executed, 31 up-to-date

Build Analyzer results available
11:38:58: Task execution finished ':app:testDebugUnitTest --tests "com.example.benchproject.app.popular.movies.viewModel.PopularMoviesViewModelTest"'.

```

**Comment by BorysChajdas** [comment]
You have created a mock for the navigator but don't specify what should happen after calling `errorSnackBar` that's why the exception is raised. You can use `relaxed` mock (probably `relaxUnitFun` will do the job) or specify that calling `errorSnackBar` should just run `every { navigator.errorSnackbar(...) } just Runs`

**Comment by Nataniel-Antosik** [comment]
When I added
`every { popularMoviesFragmentNavigator.errorSnackBar("Something went wrong") } just Runs`
everything work correct thanks :D

File: app/src/test/java/com/example/benchproject/test/common/InstantExecutorExtension.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+    override fun beforeEach(context: ExtensionContext?) {
+        ArchTaskExecutor.getInstance()
+            .setDelegate(object : TaskExecutor() {
+                override fun executeOnDiskIO(runnable: Runnable) = runnable.run()
+
+                override fun postToMainThread(runnable: Runnable) = runnable.run()
+
+                override fun isMainThread(): Boolean = true
+            })
```
Comment (BorysChajdas) [inline app/src/test/java/com/example/benchproject/test/common/InstantExecutorExtension.kt]: Minor: Wrap `)` to new line

File: app/src/test/java/com/example/benchproject/app/popular/movies/viewModel/PopularMoviesViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    private val badResponseFromRepository =
+        Result.failure<List<MovieModel>>(
+            Throwable("Unable to resolve host \"api.themoviedb.org\": No address associated with hostname")
+        )
+    private val getPopularMoviesUseCase: GetPopularMoviesUseCase = mockk()
+    private val popularMoviesFragmentNavigator: PopularMoviesFragmentNavigator = mockk()
+
+    @BeforeAll
+    fun initAll() {
+        Dispatchers.setMain(Dispatchers.Unconfined)
```
Comment (BorysChajdas) [inline app/src/test/java/com/example/benchproject/app/popular/movies/viewModel/PopularMoviesViewModelTest.kt]: Add it to base class `InstantExecutorExtension`

File: app/src/test/java/com/example/benchproject/app/popular/movies/viewModel/PopularMoviesViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @AfterAll
+    fun tearDown() {
+        Dispatchers.resetMain()
```
Comment (BorysChajdas) [inline app/src/test/java/com/example/benchproject/app/popular/movies/viewModel/PopularMoviesViewModelTest.kt]: Add it to base class `InstantExecutorExtension`

File: app/src/test/java/com/example/benchproject/test/common/InstantExecutorExtension.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package com.example.benchproject.test.common
+
+import androidx.arch.core.executor.ArchTaskExecutor
+import androidx.arch.core.executor.TaskExecutor
+import org.junit.jupiter.api.extension.AfterEachCallback
+import org.junit.jupiter.api.extension.BeforeEachCallback
+import org.junit.jupiter.api.extension.ExtensionContext
+
+class InstantExecutorExtension : BeforeEachCallback, AfterEachCallback {
```
Comment (BorysChajdas) [inline app/src/test/java/com/example/benchproject/test/common/InstantExecutorExtension.kt]: Could be renamed to more meaningful name like `LiveDataTest`

**Review by BorysChajdas — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/example/benchproject/test/common/InstantExecutorExtension.kt (hunk 4 above)
Comment (Nataniel-Antosik) [reply app/src/test/java/com/example/benchproject/test/common/InstantExecutorExtension.kt]: I renamed this class on `LiveDataTest`

File: app/src/test/java/com/example/benchproject/app/popular/movies/viewModel/PopularMoviesViewModelTest.kt (hunk 3 above)
Comment (Nataniel-Antosik) [reply app/src/test/java/com/example/benchproject/app/popular/movies/viewModel/PopularMoviesViewModelTest.kt]: I added it to this class

File: app/src/test/java/com/example/benchproject/app/popular/movies/viewModel/PopularMoviesViewModelTest.kt (hunk 2 above)
Comment (Nataniel-Antosik) [reply app/src/test/java/com/example/benchproject/app/popular/movies/viewModel/PopularMoviesViewModelTest.kt]: I added it to this class

File: app/src/test/java/com/example/benchproject/test/common/InstantExecutorExtension.kt (hunk 1 above)
Comment (Nataniel-Antosik) [reply app/src/test/java/com/example/benchproject/test/common/InstantExecutorExtension.kt]: I added to new line

File: app/src/test/java/com/example/benchproject/test/common/LiveDataTest.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        Dispatchers.setMain(Dispatchers.Unconfined)
+        ArchTaskExecutor.getInstance()
+            .setDelegate(object : TaskExecutor() {
+                override fun executeOnDiskIO(runnable: Runnable) = runnable.run()
+
+                override fun postToMainThread(runnable: Runnable) = runnable.run()
+
+                override fun isMainThread(): Boolean = true
+            }
+            )
```
Comment (BorysChajdas) [inline app/src/test/java/com/example/benchproject/test/common/LiveDataTest.kt:25]: Bad formatting, I don't know why ktlint has not failed

Comment (Nataniel-Antosik) [reply app/src/test/java/com/example/benchproject/test/common/LiveDataTest.kt:25]: Maybe now looks better? :D

Comment (BorysChajdas) [reply app/src/test/java/com/example/benchproject/test/common/LiveDataTest.kt:25]: Probably such thing is not an error for ktlint, but it should look like this:
```
    override fun beforeEach(context: ExtensionContext?) {
        Dispatchers.setMain(Dispatchers.Unconfined)
        ArchTaskExecutor.getInstance()
            .setDelegate(
                object : TaskExecutor() {
                    override fun executeOnDiskIO(runnable: Runnable) = runnable.run()

                    override fun postToMainThread(runnable: Runnable) = runnable.run()

                    override fun isMainThread(): Boolean = true
                }
            )
    }
```

**Review by BorysChajdas — APPROVED** [review]

File: app/src/test/java/com/example/benchproject/test/common/LiveDataTest.kt (hunk 5 above)
Comment (Nataniel-Antosik) [reply app/src/test/java/com/example/benchproject/test/common/LiveDataTest.kt:25]: Yes, I changed the code and looks the same

**Pull request merged**